pkg/build/generator/generator.go

```diff
@@ -575,25 +575,54 @@ func (g *BuildGenerator) resolveImageStreamReference(ctx kapi.Context, from kapi
 	glog.V(4).Infof("Resolving ImageStreamReference %s of Kind %s in namespace %s", from.Name, from.Kind, namespace)
 	switch from.Kind {
 	case "ImageStreamImage":
-		imageStreamImage, err := g.Client.GetImageStreamImage(kapi.WithNamespace(ctx, namespace), from.Name)
+		name, id, err := imageapi.ParseImageStreamImageName(from.Name)
 		if err != nil {
 			err = resolveError(from.Kind, namespace, from.Name, err)
 			glog.V(2).Info(err)
 			return "", err
 		}
-		image := imageStreamImage.Image
-		glog.V(4).Infof("Resolved ImageStreamReference %s to image %s with reference %s in namespace %s", from.Name, image.Name, image.DockerImageReference, namespace)
-		return image.DockerImageReference, nil
+		stream, err := g.Client.GetImageStream(kapi.WithNamespace(ctx, namespace), name)
+		if err != nil {
+			err = resolveError(from.Kind, namespace, from.Name, err)
+			glog.V(2).Info(err)
+			return "", err
+		}
+		tag, _ := imageapi.LatestImageTagEvent(stream, id)
+		if len(tag) == 0 {
+			err = resolveError(from.Kind, namespace, from.Name, fmt.Errorf("unable to find tag associated with image %q", id))
+			glog.V(2).Info(err)
+			return "", err
+		}
+		reference, ok := imageapi.ResolveLatestTaggedImage(stream, tag)
```

Comment:
This is wrong. Returned reference may be for completely different image than requested.

Imagine image stream `is` with a tag having two images in history:
```
"latest": { "items": [ "image": "a", "image": "b" ]}
```

Now this is called with a reference to image stream tag `is:b`. The `imageapi.LatestImageTagEvent` returns the right tag event with `"image": "b"`. However, this call takes the tag, looks into `is` and takes the first image `a`. So the returned reference contains image name `a` instead of `b`.